### ISSUE TITLE
Sphinx failes to document members/properties with type annotation and #: comments

### ISSUE THREAD

**Issue by orenbenkiki** [sphinx-doc/sphinx#3864]
### Problem

Sphinx doesn't fully support variable type annotations.
Specifically, in ``pycode/Grammar-py3.txt``, the BNF rule:
```
expr_stmt: testlist_star_expr (augassign (yield_expr|testlist) |
                     ('=' (yield_expr|testlist_star_expr))*)
```

Does not allow for a type annotation. To do so, it should probably have said:
```
expr_stmt: testlist_star_expr (augassign (yield_expr|testlist) |
                     ([':' test] '=' (yield_expr|testlist_star_expr))*)
```
I'm not certain why ``test`` is used for parsing type annotations, but it is used in ``tfpdef: NAME [':' test]`` so I _assume_ that's the right thing... - at any rate, it _seems_ to work.

#### Procedure to reproduce the problem
```
#: Is documented.
is_documented = 1

#: Not documented :-(
not_documented: Optional[int] = 2
```

This is using ``sphinx.ext.autodoc``, of course. The same problem will happen when trying to document members ``self.not_documented: SomeType = some_value``.

#### Error logs / results

There are no errors, however the ``not_documented`` variable is silently ignored and not shown in the documentation.

#### Expected results

At minimum: the ``not_documented`` variable should be shown in the generated documentation.

Ideally: the documentation should include the type annotation. However, I realize Sphinx does not use type annotations "out of the box", requiring the use of extensions (I use a locally-patched version of ``autodoc_annotation``). So this might be provided as an extension (or an enhancement of an existing extension) for dealing with such annotations.

This presumes it is possible for such an extension to access the annotations (using ``inspect``?) - I'm not certain this is the case for type annotations used when assigning to ``self.member``. If such annotations are _not_ easily accessible, perhaps that would tip the scales towards ``autodoc`` embedding the annotations without requiring additional extensions...?

### Environment info
- OS: Linux (CentOS release 6.9 (Final) - don't judge me, it is a corporate environment...)
- Python version: 3.6.0
- Sphinx version: 1.6.2

**Comment by tk0miya** [comment]
I tried to fix this with updating grammar file, but I failed it.
We need to work hard to fix this.

**Comment by tk0miya** [comment]
Fixed by #3937
